packages/bundle-source/cache.js

```diff
@@ -81,57 +86,91 @@ export const makeBundleCache = (wr, cwd, readPowers, opts) => {
           modTimeByPath.set(ref, mtime);
           // console.log({ loc, mtime, ref });
         } catch (oops) {
-          console.error(oops);
+          log(oops);
         }
       }
       return readPowers.read(loc);
     };
-    const bundle = await bundleSource(
-      rootPath,
-      {},
-      { ...readPowers, read: loggedRead },
-    );
-
-    const { moduleFormat } = bundle;
-    assert.equal(moduleFormat, 'endoZipBase64');
 
-    const code = encodeBundle(bundle);
     await wr.mkdir({ recursive: true });
-    const bundleFileName = toBundleName(targetName);
-    const bundleWr = wr.neighbor(bundleFileName);
-    await bundleWr.writeText(code);
-    /** @type {import('fs').Stats} */
-    const { mtime: bundleTime } = await bundleWr.readOnly().stat();
 
-    /** @type {BundleMeta} */
-    const meta = {
-      bundleFileName,
-      bundleTime: bundleTime.toISOString(),
-      moduleSource: {
-        relative: bundleWr.readOnly().relative(srcRd.absolute()),
-        absolute: srcRd.absolute(),
-      },
-      contents: [...modTimeByPath.entries()].map(([relativePath, mtime]) => ({
-        relativePath,
-        mtime: mtime.toISOString(),
-      })),
-    };
+    const lockWr = wr.neighbor(toBundleLock(targetName));
 
-    await wr
-      .neighbor(toBundleMeta(targetName))
-      .writeText(JSON.stringify(meta, null, 2));
-    return meta;
+    // Check the bundle/meta file write lock.
+    try {
+      await lockWr.writeText('', { flag: 'wx' });
+    } catch (oops) {
+      if (oops.code !== 'EEXIST') {
+        throw oops;
+      }
+      // The lock exists, so something is already writing the bundle.
+      // eslint-disable-next-line no-use-before-define
+      return validate(targetName, rootPath);
```

Comment:
so... we win if what we're trying to write is already there?

Reply:
Clarified with a better comment:

```js
      // The lock exists, so something is already writing the bundle on our
      // behalf.
      //
      // All we need to do is try validating the bundle, which will first wait
      // for the lock to disappear before reading the freshly-written bundle.
```